docs/reference/modules/discovery.asciidoc

```diff
@@ -1,8 +1,397 @@
+that uses the https://github.com/aws/aws-sdk-java[AWS API] to find a list of seed nodes.
+
+[float]
+[[azure-classic-hosts-provider]]
+===== Azure Classic hosts provider
+
+The {plugins}/discovery-azure-classic.html[Azure Classic discovery plugin] adds a hosts provider
+that uses the Azure Classic API find a list of seed nodes.
+
+[float]
+[[gce-hosts-provider]]
+===== Google Compute Engine hosts provider
+
+The {plugins}/discovery-gce.html[GCE discovery plugin] adds a hosts provider
+that uses the GCE API find a list of seed nodes.
+
+[float]
+==== Discovery settings
+
+Discovery operates in two phases: First, each node "probes" the addresses of
+all known nodes by connecting to each address and attempting to identify the
+node to which it is connected. Secondly it shares with the remote node a list
+of all of its peers and the remote node responds with _its_ peers in turn. The
+node then probes all the new nodes about which it just discovered, requests
+their peers, and so on, until it has discovered an elected master node or
+enough other masterless nodes that it can perform an election. If neither of
+these occur quickly enough then it tries again. This process is controlled by
+the following settings.
+
+`discovery.probe.connect_timeout`::
+
+    Sets how long to wait when attempting to connect to each address. Defaults
+    to `3s`.
+
+`discovery.probe.handshake_timeout`::
+
+    Sets how long to wait when attempting to identify the remote node via a
+    handshake. Defaults to `1s`.
+
+`discovery.find_peers_interval`::
+
+    Sets how long a node will wait before attempting another discovery round.
+
+`discovery.request_peers_timeout`::
+
+    Sets how long a node will wait after asking its peers again before
+    considering the request to have failed.
+
+
+[float]
+[[modules-discovery-bootstrap-cluster]]
+=== Bootstrapping a cluster
+
+Starting an Elasticsearch cluster for the very first time requires a
+cluster bootstrapping step.
+
+The simplest way to bootstrap a cluster is by specifying the node names
+or transport addresses of at least a non-empty subset of the master-eligible nodes
+before start-up. The node setting `cluster.initial_master_nodes`, which
+takes a list of node names or transport addresses, can be either specified
+on the command line when starting up the nodes, or be added to the node
+configuration file `elasticsearch.yml`.
+
+For a cluster with 3 master-eligible nodes (named master-a, master-b, and master-c)
+the configuration will look as follows:
+
+[source,yaml]
+--------------------------------------------------
+cluster.initial_master_nodes:
+  - master-a
+  - master-b
+  - master-c
+--------------------------------------------------
+
+TODO provide another example with ip addresses (+ possibly port)
+
+Note that if you have not explicitly configured a node name, this
+name defaults to the host name, so using the host names will work as well.
```

Comment:
So host names will only work if node names haven't been explicitly configured or is it always possible to specify hostnames in the list?

Reply:
I think this sentence isn't quite right. It is always possible to specify hostnames in `cluster.initial_master_nodes`:

https://github.com/elastic/elasticsearch/blob/986bf52d1f3a3dc38ee4a29e1c8b3f5c7ef55126/server/src/main/java/org/elasticsearch/action/admin/cluster/bootstrap/TransportGetDiscoveredNodesAction.java#L135-L139

Reply:
Hmm, I'm not sure what I just said is quite right, because `discoveryNode.getAddress()` needs to know a hostname and not just the address for this to work. Needs more investigation. It might be something like "hostname works if it's the name you use in the node's publish address" perhaps.

Reply:
@dliappis I dug into this some more and discovered that today we support IP addresses and node names but not hostnames (although the node name is set to the hostname by default, so in that situation the hostname is technically ok).

Reply:
@DaveCTurner Just a confirmation from my side too that IP addresses work fine, I've also switched to this for our Rally generated configurations for Elasticsearch in https://github.com/elastic/rally-teams/blob/6acc645df9d32aad18b610b710bc54e2f73d9468/cars/v1/vanilla/templates/config/elasticsearch.yml#L74